src/qhttp/Request.cc

```diff
@@ -29,65 +29,124 @@
 // Third-party headers
 #include "boost/regex.hpp"
 
+// Local headers
+#include "lsst/log/Log.h"
+#include "qhttp/LogHelpers.h"
+
 namespace ip = boost::asio::ip;
 
+namespace {
+    LOG_LOGGER _log = LOG_GET("lsst.qserv.qhttp");
+}
+
 namespace lsst {
 namespace qserv {
 namespace qhttp {
 
-Request::Request(std::shared_ptr<ip::tcp::socket> socket)
+Request::Request(std::shared_ptr<Server> server, std::shared_ptr<ip::tcp::socket> socket)
 :
     content(&_requestbuf),
+    _server(server),
     _socket(socket)
 {
     boost::system::error_code ignore;
     localAddr = _socket->local_endpoint(ignore);
 }
 
 
-void Request::_parseHeader()
+bool Request::_parseHeader()
 {
     std::string line;
-    static boost::regex reqRe{R"(^([^ \r]+) ([^ \r]+) ([^ \r]+)\r$)"}; // e.g. "method target version"
+    if (!getline(content, line)) return false;
+
+    // Regexp to parse request line into "method target version".  Allowed character classes here are
+    // extracted from ABNF in RFC7230 and RFC3986.
+
+    static boost::regex reqRe{
+        R"(^([!#$%&'*+-.^_`|~[:digit:][:alpha:]]+) )"
+        R"(([-._-~%!$&'()*+,;=:@?/[:digit:][:alpha:]]+) )"
+        R"((HTTP/[[:digit:]]\.[[:digit:]])\r$)"
+    };
+
     boost::smatch reqMatch;
-    if (getline(content, line) && boost::regex_match(line, reqMatch, reqRe)) {
-        method = reqMatch[1];
-        target = reqMatch[2];
-        version = reqMatch[3];
-        static boost::regex headerRe{R"(^([^:\r]+): ?([^\r]*)\r$)"}; // e.g. "header: value"
-        boost::smatch headerMatch;
-        while(getline(content, line) && boost::regex_match(line, headerMatch, headerRe)) {
-            header[headerMatch[1]] = headerMatch[2];
+    if (!boost::regex_match(line, reqMatch, reqRe)) {
+        LOGLS_WARN(_log, logger(_server) << logger(_socket) << "bad start line: " << ctrlquote(line));
+        return false;
+    }
+
+    method = reqMatch[1];
+    target = reqMatch[2];
+    version = reqMatch[3];
+
+    // Regexp to parse header field lines into "header: value".  Allowed character classes here are
+    // extracted from ABNF in RFC7230.
+
+    static boost::regex headerRe{
+        R"(^([!#$%&'*+-.^_`|~[:digit:][:alpha:]]+))"
+        R"(:[ \t]*)"
+        R"(([\x20-\x7e \t]*?))"
+        R"([ \t]*\r$)"
+    };
+
+    boost::smatch headerMatch;
+
+    while(getline(content, line)) {
+        if (!boost::regex_match(line, headerMatch, headerRe)) {
+            if (line == "\r") break;
```

Comment:
Why not check for this condition before matching the regular expression? My understanding of the protocol is that the empty line indicates the end of the header (and the header may not have any lines at all).

Reply:
`"\r"` line on its own, in the no-headers case, will still match the regexp.  However, I think your suggest is still good for code clarity, so will do.